### ISSUE TITLE
Mass Moves

### ISSUE THREAD

**Issue by LazerRhome** [tmanifold/mvbot#4]
Just seeing if it's possible to make it so that we can move multiple messages at once - maybe if we make it so you can post multiple IDs separated by a comma? (don't know if that's possible as I'm not a coder!)

Thank you!

**Comment by WeserBlitz** [comment]
> Just seeing if it's possible to make it so that we can move multiple messages at once - maybe if we make it so you can post multiple IDs separated by a comma? (don't know if that's possible as I'm not a coder!)
> 
> Thank you!

I'm really happy with the bot, was a headache bothering me for some time that i couldn't move messages between channels. BUT: I would ADORE the developer if batch- moving messages could be enabled! Other than that, happy new year and keep up the good work!

**Comment by tmanifold** [comment]
Thanks for the feedback! I will look into this when I have some extra time.

**Comment by WeserBlitz** [comment]
> Thanks for the feedback! I will look into this when I have some extra time.

You rock! Thank you a million!!! Hero!

**Comment by videovillain** [comment]
I too would love some batch movement capabilities!! :D

Maybe allowing something like this:
!mv <message-id,message-id,message-id,message-id,message-id> <target-channel> ["reason"]

Or maybe somehow select all messages from last (15 minutes) -for when people go off on a tangent- to a new channel:
!mv <all> <15> <target-channel> ["reason"]

And of course, hopefully they would be in the same order as before the move :D

Thanks for your hard work!

**Comment by tmanifold** [comment]
Due to all the interest this feature is actually in the works. I recently did rewrite of large portions of the code to allow for this functionality and am playing with some different methods of implementation. This should be live in some form or another in the next update.

**Comment by WeserBlitz** [comment]
> Due to all the interest this feature is actually in the works. I recently did rewrite of large portions of the code to allow for this functionality and am playing with some different methods of implementation. This should be live in some form or another in the next update.

This is great news! @tmanifold you're awesome <3 Thanks!

**Comment by tmanifold** [comment]
The new version will make use of command-line style switches. Currently, I've implemented a switch to move the next `n` number of messages in addition to the specified message. I am thinking about ways to implement a delimited list. I also like @videovillain's idea of moving messages within the last `x` amount of time, so i will probably work on this too. 

An example of the new syntax might be: `!mv -m <message-id> -d <destination-channel> [-c "comment"] [-n <number-of-messages]`

I'm planning to make the default behavior for bulk moves to only write the mvbot header (the part containing the original user, channel, time, and comment) for the first moved message for each user in a span of messages:
![image](https://user-images.githubusercontent.com/11261953/112763058-e4d3fe00-8fd0-11eb-9c73-94307a3711eb.png)




I am planning to make delimited lists an exception to this rule; they will have a header for every message. This can be overridden with another planned switch, ex `--no-header`.

Additionally, if i can maintain backwards compatibility with the current command syntax then I will. However it is likely to be deprecated in future versions.

The current development build is on the `dev` branch, and implements the `-n` switch.

I appreciate everyone's feedback and input. This type of engagement helps make mvbot better!

**Comment by changeling** [comment]
So the no-header option doen'st exist for the current master, is that correct? Is there a rough timeline for the next update, or should I shut up and sit down to wait patiently. :)

**Comment by changeling** [comment]
Will the next update allow moving a message cleanly, so that it appears just as it did before the move, if the no-headers flag is set? That's really what I'm looking for, those this is sooo close. :)

**Comment by tmanifold** [comment]
@changeling correct, these features are not yet live. My plan for something like a --no-header flag is to simply have the bot move the message content and not post extra info about who or when the original message occurred. This will likely be detailed greater on the dev branch when I get a more functional build ready.

As for a timeline on the next update, I don't have much of an estimate. I am working on my final year of university, and work on the bot when time and energy allows. 

You patience is appreciated, big changes are to come and I think they will be good for the functionality and utility of mvbot.

**Comment by changeling** [comment]
Thanks for all your work on this! Absolutely patient on my end.

Just for clarity, will the moved messages appear just as it did in the
original channel, eventually? Date, time, poster? Ideally we’d like to have
position and threading appear the same in the new location.


And good travels with university!

Chris

**Comment by tmanifold** [comment]
> Just for clarity, will the moved messages appear just as it did in the original channel, eventually? Date, time, poster?

Unfortunately this isn't possible due to the limitations of the Discord API. The mvbot header is a stand-in for this.

**Comment by jerbear2008** [comment]
One idea I had would be to use webhooks to post the messages with a custom name and avatar. However, a webhook would need to be made for every channel.

**Comment by tmanifold** [comment]
> One idea I had would be to use webhooks to post the messages with a custom name and avatar. However, a webhook would need to be made for every channel.

The way I understand webhooks, there isn't really a feasible way to implement this. Unless the discord API significantly changes, the bot copying message content and referencing the original user is likely the best that can be done.

**Comment by changeling** [comment]
> > Just for clarity, will the moved messages appear just as it did in the original channel, eventually? Date, time, poster?
> 
> Unfortunately this isn't possible due to the limitations of the Discord API. The mvbot header is a stand-in for this.

Gotcha! Sorry for the delayed response.

On that note, though, will the resulting messages sans header at least look "clean?" That is, will they simply be the content of the message?

And the obligatory, no pressure just asking, question:

Any idea of timeframe this might commit?

**Comment by jerbear2008** [comment]
Yeah, it would do that. However, you would miss timestamps and where the message originally came from.

**Comment by tmanifold** [comment]
> And the obligatory, no pressure just asking, question:
> 
> Any idea of timeframe this might commit?

@changeling, please refer to [my earlier reply ](https://github.com/tmanifold/mvbot/issues/4#issuecomment-816929899)

**Comment by changeling** [comment]
> > And the obligatory, no pressure just asking, question:
> > Any idea of timeframe this might commit?
> 
> @changeling, please refer to [my earlier reply ](https://github.com/tmanifold/mvbot/issues/4#issuecomment-816929899)

Which I should have done before. Thanks for the pointer, with apologies.

**Comment by zkll505** [comment]
I'm gonna be honest, I love this bot, and I wish I could implement this command into my own bot, but I suck at coding, since I just started out a couple months ago. I tried re-coding it to work within an "advanced" command handler, to make it work in a bot with multiple commands, using the multiple file separation method, and was stumped within 20 minutes.

**Comment by tmanifold** [comment]
> I'm gonna be honest, I love this bot, and I wish I could implement this command into my own bot, but I suck at coding, since I just started out a couple months ago. I tried re-coding it to work within an "advanced" command handler, to make it work in a bot with multiple commands, using the multiple file separation method, and was stumped within 20 minutes.

Glad you like the bot! I have a working version of this in the current dev branch. Just a few things to iron out when I have time and hopefully it will be ready to push  to main. As for the coding, don't give up! Just like any other skill, it just takes time and practice.

**Comment by DorkyDino** [comment]
Excited to see this implemented! I'm currently trying to combine a few channels and this would make it so much easier! Thanks for doing what Discord has refused to do for years. lol
